### PR TITLE
Add hardhat_metadata RPC method

### PR DESCRIPTION
**Pull request by fvictorio**
Adds a new `hardhat_metadata` RPC method. This method is meant to be used by other tools, like wallets or deployment systems, to get information about the instance of the network.

### CODE REVIEW

File: packages/hardhat-core/src/internal/hardhat-network/provider/modules/web3.ts [hunk 1]
```diff
@@ -32,9 +34,7 @@ export class Web3Module {
   }
 
   private async _clientVersionAction(): Promise<string> {
-    const hardhatPackage = await getPackageJson();
-    const ethereumjsVMPackage = require("@nomicfoundation/ethereumjs-vm/package.json");
-    return `HardhatNetwork/${hardhatPackage.version}/@nomicfoundation/ethereumjs-vm/${ethereumjsVMPackage.version}`;
+    return this._node.getClientVersion();
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/hardhat-network/provider/modules/web3.ts:37]: The value returned in the `clientVersion` field of the metadata is the same one that is returned by `web3_clientVersion`. Is that ok?

File: packages/hardhat-core/src/internal/hardhat-network/provider/utils/makeForkClient.ts [hunk 2]
```diff
@@ -88,7 +90,14 @@ Please use block number ${lastSafeBlock} or wait for the block to get ${
     cacheToDiskEnabled ? forkCachePath : undefined
   );
 
-  return { forkClient, forkBlockNumber, forkBlockTimestamp };
+  const forkBlockHash = block.hash;
+
+  assertHardhatInvariant(
+    forkBlockHash !== null,
+    "Forked block should have a hash"
+  );
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/hardhat-network/provider/utils/makeForkClient.ts:98]: I think only pending blocks have a null hash, so this should be fine.

File: packages/hardhat-core/src/internal/core/jsonrpc/types/output/metadata.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  // The latest block's number in Hardhat Network
+  latestBlockNumber: number;
+
+  // The latest block's hash in Hardhat Network
+  latestBlockHash: string;
+
+  // This field is only present when Hardhat Network is forking another chain.
+  forkedNetwork?: {
+    // The chainId of the network that is being forked
+    chainId: number;
```
Comment (alcuadrado) [inline packages/hardhat-core/src/internal/core/jsonrpc/types/output/metadata.ts:27]: Should add `networkId` here, for consistency?

File: packages/hardhat-core/src/internal/hardhat-network/provider/modules/web3.ts (hunk 1 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/hardhat-network/provider/modules/web3.ts:37]: Yep

File: packages/hardhat-core/src/internal/hardhat-network/provider/utils/makeForkClient.ts (hunk 2 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/hardhat-network/provider/utils/makeForkClient.ts:98]: Yes, this is a valid assumption

**Review by alcuadrado — APPROVED** [review]
LGTM.

I just left a comment about fork network id, mostly to make the API more consistent.

File: packages/hardhat-core/src/internal/core/jsonrpc/types/output/metadata.ts (hunk 3 above)
Comment (fvictorio) [reply packages/hardhat-core/src/internal/core/jsonrpc/types/output/metadata.ts:24]: Removed `networkId` instead.

**Comment by alcuadrado** [comment]
> I just left a comment about fork network id, mostly to make the API more consistent.

We decided to drop `networkId` instead, as it's not clearly why one would use it for. We can reintroduce it if needed.

**Pull request merged**